### PR TITLE
Adding infrastructure for normalizing SyGuS grammars

### PR DESCRIPTION
**Pull request by HanielB**
This is the first step towards addressing #1304. This pull request:

- adds an option for normalizing user provided and automatically generated SyGuS grammars
- adds a new module for normalizing SyGuS grammars
- adds the interface of the module that reads a TypeNode and builds a new TypeNode equivalent to it. For now no transformation is performed, the TypeNodes are just reconstructed.
			
The point here is to be sure that the datatypes and typenodes are correctly reconstructed (which is tricky). Now normalization can begin.

### CODE REVIEW

File: src/theory/quantifiers/sygus_grammar_simp.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
+/*********************                                                        */
+/*! \file sygus_grammar_cons.cpp
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.cpp]: Wrong file name

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+   *
+   * This contains global information about the synthesis conjecture. */
+  CegConjecture* d_parent;
+
+  /** sygus term database associated with this utility */
+  TermDbSygus* d_tds;
+
+  /** normalize integer type
+   *
+   * TODO actually perform the normalization
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: reference #1304 here

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+   *
+   * ind is the indice of the analyzed typeobject in tos
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: indice -> index

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+   * new types created during normalization will be added to tos and
+   * tn_to_unres
+   *
+   * sygus_vars is used as above for datatype construction */
+  void normalizeSygusInt(unsigned ind,
+                         std::vector<TypeObject>& tos,
+                         std::map<TypeNode, Type>& tn_to_unres,
+                         Node sygus_vars);
+
+  /** Traverses the datatype representing a typenode and collects the types it
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: Be explicit here, src_tn the DatatypeType we are traversing I assume?

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+   * contains
+   *
+   * for each new typenode a TypeObject is created, with an unresolved type and
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: Capitalize

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+{
+  /* Both constructors create an unresolved type and datatype with the given
+   * name. An original typenode is optional, since normalized types can be
+   * created from scratch during normalization */
+  TypeObject(TypeNode src_tn, std::string type_name);
+  TypeObject(std::string type_name);
+  ~TypeObject() {}
+
+  /* The original typenode this TypeObject is normalizing */
+  TypeNode tn;
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: All of these need "d_" prefixes for data members.

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+class CegConjecture;
+
+/** Keeps the necessary information in building a normalized type:
+ *
+ * the original typenode, from which the datatype representation can be
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: This sentence is hard to follow.

File: src/theory/quantifiers/sygus_grammar_simp.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+   * created from scratch during normalization */
+  TypeObject(TypeNode src_tn, std::string type_name);
+  TypeObject(std::string type_name);
+  ~TypeObject() {}
+
+  /* The original typenode this TypeObject is normalizing */
+  TypeNode tn;
+  /* Operators for each construct. */
+  std::vector<Expr> ops;
+  /* Names for each construct. */
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_simp.h]: construct -> constructor

File: src/theory/quantifiers/sygus_grammar_norm.cpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  /* Make int_type and zero */
+  NodeManager* nm = NodeManager::currentNM();
+  /* Accumulates all typing information for normalization and reconstruction */
+  std::vector<TypeObject> tos;
+  /* Allows retrieving respective unresolved types for the sygus types of the
+   * original type nodes */
+  std::map<TypeNode, Type> tn_to_unres;
+  std::map<TypeNode, bool> visited;
+  collectInfoFor(tn, tos, tn_to_unres, visited);
+  /* Build datatypes TODO and normalize accordingly */
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_norm.cpp]: reference #1304 here too :)

File: src/theory/quantifiers/sygus_grammar_norm.h [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "theory/quantifiers_engine.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+class CegConjecture;
+
+/** Keeps the necessary information in building a normalized type:
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_norm.h]: I would just say "information about a sygus type".  Itisn't necessarily tied to the normalization, right?

File: src/theory/quantifiers/sygus_grammar_norm.h [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+struct TypeObject
+{
+  /* Both constructors create an unresolved type and datatype with the given
+   * name. An original typenode is optional, since normalized types can be
+   * created from scratch during normalization */
+  TypeObject(TypeNode src_tn, std::string type_name);
+  TypeObject(std::string type_name);
+  ~TypeObject() {}
+
+  /* The original typenode this TypeObject is normalizing */
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus_grammar_norm.h]: TypeObject isn't doing the normalization, right?

File: src/theory/quantifiers/sygus_grammar_norm.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief implementation of class for for simplifying SyGuS grammars after they
+ ** are encoded into datatypes.
+ **/
+
+#include "theory/quantifiers/sygus_grammar_norm.h"
+
+#include <stack>
```
Comment (aniemetz) [inline src/theory/quantifiers/sygus_grammar_norm.cpp]: This include is unused.

File: src/theory/quantifiers/sygus_grammar_norm.h [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+class CegConjecture;
+
+/** Keeps the necessary information in building a normalized type:
+ *
+ * the original typenode, from which the datatype representation can be
+ * extracted
+ * the operators, names and argument types for each constructor
+ * the unresolved type used as placeholder for references of the yet to be built
+ * type
+ * a datatype to represent the structure of the typenode for the new type */
```
Comment (aniemetz) [inline src/theory/quantifiers/sygus_grammar_norm.h]: Maybe make it more obvious that this is a list?
